xdsl/transforms/stencil_shape_minimize.py

```diff
@@ -42,6 +50,23 @@ def match_and_rewrite(self, op: func.FuncOp, rewriter: PatternRewriter, /):
             op.update_function_type()
 
 
+@dataclass
+class RestrictStoreOp(RewritePattern):
+    restrict: tuple[int, ...]
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: stencil.StoreOp, rewriter: PatternRewriter, /):
+        if self.restrict:
+            new_bounds = [
+                (min(lower_bound, bound_lim), min(upper_bound, bound_lim))
+                for lower_bound, upper_bound, bound_lim in zip(
+                    op.bounds.lb, op.bounds.ub, self.restrict
+                )
+            ]
+            new_bounds_attr = stencil.StencilBoundsAttr(new_bounds)
+            self.bounds = new_bounds_attr
```

Comment:
Assigning to `op.bounds` would actually mean the changes pass the rewriter. It may be safer to rebuild the op via `rewriter.replace_matched_op(stencil.StoreOp.get(...))`.

To prevent this from infinitely rewriting the op, it also needs either `apply_recursively=False` in the PatternRewriteWalker, or a check in this class whether this should be applied, for instance, by checking if `new_bounds_attr != op.bounds` to know if this needs updating, and bail out early otherwise.

Reply:
ok, thanks for that. I'll work on this now

Reply:
I think that because I now have GreedyRewritePatternApplier, I can't have apply_recursively=False.. so I will go with the other option for now but please let me know if I'm missing something simpler